### PR TITLE
Cli: commands to add books, items, loans and remove loans

### PR DESCRIPTION
**Pull request by KonstantinaStoikou**
- `cds-ils demo import-demo-data --are-docs --path <path_to_json_file>` to add demo documents
- `cds-ils demo import-demo-data --are-items --path <path_to_json_file>` to add demo items
- `cds-ils demo user-test-prepare --user-email <user_email>` to add an ongoing loan for a user that starts 3 weeks before today and a past loan that started a year before today and has ended
- `cds-ils demo user-test-clean --user-email <user_email> --given-date <date>` to delete all loans and document requests made by the given user on a given date (default date=today)

### CODE REVIEW

File: scripts/life_data_setup [hunk 1]
```diff
@@ -15,6 +19,8 @@ invenio fixtures location
 invenio fixtures vocabularies
 invenio fixtures demo-patrons
 invenio files location --default ils /tmp/ils-files
+cds-ils migration internal-locations $LIBRARIES_PATH
```
Comment (kprzerwa) [inline scripts/life_data_setup:22]: this is not needed, there is the same command below under the commands

File: cds_ils/cli.py [hunk 2]
```diff
@@ -130,6 +195,16 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+    patrons = User.query.all()
+    indexer = current_app_ils.patron_indexer
+
+    click.secho("Now indexing {0} patrons".format(len(patrons)), fg="green")
+
+    Patron = current_app_ils.patron_cls
+    for pat in patrons:
+        patron = Patron(pat.id)
+        indexer.index(patron)
+
```
Comment (kprzerwa) [inline cds_ils/cli.py:208]: you can use instead
`run command("patrons index")` 
to be checked if it works!

File: cds_ils/cli.py [hunk 3]
```diff
@@ -159,3 +234,390 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+        self.items = {"objs": []}
+        self.intlocs = {"objs": []}
+        self.loans = {"objs": []}
+
+    def pids(self, collection, pid_field):
+        """Get a list of PIDs for a collection."""
+        return [obj[pid_field] for obj in getattr(self, collection)["objs"]]
+
+
+class DocumentGenerator(Generator):
```
Comment (kprzerwa) [inline cds_ils/cli.py:256]: let's talk about it IRL

File: cds_ils/cli.py [hunk 4]
```diff
@@ -198,3 +241,290 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+    click.secho("Items were created successfully.", fg="blue")
+
+
+@demo.command()
+@click.option("--user-email", "user_email")
+@click.option("--is-past-loan", is_flag=True)
+@with_appcontext
+def create_loan(user_email, is_past_loan):
+    """Create a loan."""
+    patron = User.query.filter_by(email=user_email).one_or_none()
```
Comment (ntarocco) [inline cds_ils/cli.py:325]:
```suggestion
    patron = User.query.filter_by(email=user_email).one()
```
I think we should fail if not found...

File: cds_ils/cli.py [hunk 5]
```diff
@@ -198,3 +241,290 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+    patron_pid = patron.get_id()
+
+    loc_pid, _ = current_app_ils.get_default_location_pid
+
+    librarian_pid = (
+        User.query.filter_by(email="librarian@test.ch").one_or_none().get_id()
```
Comment (ntarocco) [inline cds_ils/cli.py:331]: this does not exist in sandbox env. Why not using the `patron_pid`?

File: cds_ils/cli.py [hunk 6]
```diff
@@ -198,3 +241,290 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+        "pickup_location_pid": "{}".format(loc_pid),
+        "transaction_location_pid": "{}".format(loc_pid),
+        "transaction_user_pid": "{}".format(librarian_pid),
+        "delivery": {"method": delivery},
+    }
+
+    if is_past_loan:
+        loan_dict["state"] = "ITEM_RETURNED"
+        loan_dict["document_pid"] = "qaywb-gfe4B"
+        transaction_date = start_date = arrow.utcnow() - timedelta(days=365)
```
Comment (ntarocco) [inline cds_ils/cli.py:350]: where is `start_date` defined?

File: cds_ils/cli.py [hunk 7]
```diff
@@ -198,3 +241,290 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+
+    loan_dict["transaction_date"] = transaction_date.isoformat()
+    loan_dict["start_date"] = start_date.date().isoformat()
+    loan_dict["end_date"] = end_date.date().isoformat()
+    loan_dict["extension_count"] = randint(0, 3)
+    loan_dict["item_pid"] = {"type": ITEM_PID_TYPE, "value": item_pid}
+
+    loan = Loan.create(loan_dict)
+    minter(CIRCULATION_LOAN_PID_TYPE, "pid", loan)
+    loan.commit()
```
Comment (ntarocco) [inline cds_ils/cli.py:368]:
```suggestion
```

File: cds_ils/cli.py [hunk 8]
```diff
@@ -198,3 +241,290 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+    db.session.commit()
```
Comment (ntarocco) [inline cds_ils/cli.py:369]:
```suggestion
```

File: cds_ils/cli.py [hunk 9]
```diff
@@ -198,3 +241,290 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+    LoanIndexer().index(loan)
+    item = Item.get_record_by_pid(item_pid)
+    ItemIndexer().index(item)
```
Comment (ntarocco) [inline cds_ils/cli.py:372]:
```suggestion
    current_app_ils.item_indexer.index(item)
```

File: cds_ils/cli.py [hunk 10]
```diff
@@ -198,3 +241,290 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+    loan_dict["start_date"] = start_date.date().isoformat()
+    loan_dict["end_date"] = end_date.date().isoformat()
+    loan_dict["extension_count"] = randint(0, 3)
+    loan_dict["item_pid"] = {"type": ITEM_PID_TYPE, "value": item_pid}
+
+    loan = Loan.create(loan_dict)
+    minter(CIRCULATION_LOAN_PID_TYPE, "pid", loan)
+    loan.commit()
+    db.session.commit()
+    LoanIndexer().index(loan)
```
Comment (ntarocco) [inline cds_ils/cli.py:370]:
```suggestion
    current_circulation.loan_indexer.index(loan)
```

File: cds_ils/cli.py [hunk 11]
```diff
@@ -198,3 +241,290 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+    )
+
+
+@demo.command()
+@click.option("--user-email", "user_email")
+@click.option("--given-date", "given_date")
+@with_appcontext
+def clean_loans(user_email, given_date):
+    """Clean loans and doc requests created by given user on given date."""
+    indexer = RecordIndexer()
```
Comment (ntarocco) [inline cds_ils/cli.py:394]: as the other comments, we should use the specific indexer, so referenced records will be re-indexed too.

File: cds_ils/cli.py [hunk 12]
```diff
@@ -198,3 +241,290 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+
+    patron = User.query.filter_by(email=user_email).one_or_none()
```
Comment (ntarocco) [inline cds_ils/cli.py:396]:
```suggestion
    patron = User.query.filter_by(email=user_email).one()
```
same as above, fail if not found

File: cds_ils/cli.py [hunk 13]
```diff
@@ -198,3 +241,290 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+    )
+
+    Loan = current_circulation.loan_record_cls
+    for hit in patron_loans:
+        loan = Loan.get_record_by_pid(hit.pid)
+        loan.delete(force=True)
+        db.session.commit()
+        LoanIndexer().delete(loan)
+        if "item_pid" in loan:
+            item = Item.get_record_by_pid(loan["item_pid"]["value"])
```
Comment (ntarocco) [inline cds_ils/cli.py:439]:
```suggestion
            item = current_app_ils.record_cls.get_record_by_pid(loan["item_pid"]["value"])
```

File: cds_ils/cli.py [hunk 14]
```diff
@@ -198,3 +241,290 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+            loan_index = current_circulation.loan_search_cls.Meta.index
+            wait_es_refresh(loan_index)
+            ItemIndexer().index(item)
```
Comment (ntarocco) [inline cds_ils/cli.py:442]: same here

**Review by ntarocco — COMMENTED** [review]
Good job!!! 👏 
Sorry for the comments, but they are all the same, about not hardcoding class names and avoid extra `db.commit`

File: cds_ils/cli.py (hunk 6 above)
Comment (KonstantinaStoikou) [reply cds_ils/cli.py:350]: In the same line with `transaction_date`, I assign the same value to both of them.

File: cds_ils/demo_data/items.json [hunk 15]
```diff
@@ -0,0 +1,68 @@
+[
+  {
+    "pid": "678e3-an678A",
```
[the remark this answers is not in the document]
Comment (KonstantinaStoikou) [reply cds_ils/demo_data/items.json:3]: yes I thought it was the easiest solution. I changed the last letter to a capital one to differentiate it with the real pids.

File: cds_ils/cli.py [hunk 16]
```diff
@@ -198,3 +241,290 @@ def run_command(command, catch_exceptions=False, verbose=True):
...
+def mint_record_pid(pid_type, pid_field, record):
+    """Mint the given PID for the given record."""
+    PersistentIdentifier.create(
+        pid_type=pid_type,
+        pid_value=record[pid_field],
+        object_type="rec",
+        object_uuid=record.id,
+        status=PIDStatus.REGISTERED,
+    )
+    db.session.commit()
```
[the remark this answers is not in the document]
Comment (KonstantinaStoikou) [reply cds_ils/cli.py:245]: it didn't work without it so I kept it in the code

File: cds_ils/cli.py (hunk 8 above)
Comment (KonstantinaStoikou) [reply cds_ils/cli.py:341]: it didn't work without it

File: cds_ils/cli.py [hunk 17]
```diff
@@ -87,10 +115,10 @@ def location():
         "saturday",
         "sunday",
     ]
-
     closed = ["saturday", "sunday"]
     times = [
-        {"start_time": "08:30", "end_time": "18:00"},
+        {"start_time": "08:00", "end_time": "12:00"},
```
[the remark this answers is not in the document]
Comment (kprzerwa) [reply cds_ils/cli.py]: if you rebased this part is already done

File: cds_ils/cli.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (kprzerwa) [reply cds_ils/cli.py:245]: I confirm, I saw it not working without this line, we didn't investigate why yet. I will transfer this to a note to be checked when someone is free

File: cds_ils/cli.py [hunk 18]
```diff
@@ -102,21 +118,32 @@ def location():
                 **({"times": times} if is_open else {}),
             }
         )
-
+    last_date = date.today()
+    opening_exceptions = []
+    for i in range(randint(0, 3)):
```
Comment (kprzerwa) [inline cds_ils/cli.py:123]: please remove the exceptions, since we are moving to the real life data

**Review by kprzerwa — APPROVED** [review]

**Pull request merged**